File: problem81.cpp
#include <unistd.h>
#include <stdlib.h>
#include <iostream>
#include <fstream>
using namespace std;

const int gridSize = 80;

unsigned int find_min_path(int grid[gridSize][gridSize]);
unsigned int find_min_path_helper(int grid[gridSize][gridSize], int x, int y, unsigned int currentSum);
unsigned int myMin(unsigned int num1, unsigned int num2);

int main()
{
	ifstream matrixFile;
	int i;
	int j;
	
	int grid[gridSize][gridSize];
	
	matrixFile.open("p081_matrix.txt");
	
	for(i=0;i<gridSize;i++)
	{
		char temp;
		
		for(j=0;j<gridSize;j++)
		{
			matrixFile >> grid[i][j];
			matrixFile >> temp;
		}
	}
	
	matrixFile.close();
	
	/*
	for(i=0;i<gridSize;i++)
	{
		for(j=0;j<gridSize;j++)
		{
			cout << grid[i][j] << " ";
		}
		cout << endl;
	}
	*/
	
	cout << "Minimum Path: " << find_min_path(grid) << endl;
}

unsigned int find_min_path(int grid[gridSize][gridSize])
{
	return find_min_path_helper(grid, 0, 0, 0);
}

unsigned int find_min_path_helper(int grid[gridSize][gridSize], int x, int y, unsigned int currentSum)
{
	currentSum += grid[x][y];
	
	if(x == gridSize-1 && y == gridSize-1)
		return currentSum;
	//row
	else if(x == gridSize-1)
		return currentSum + find_min_path_helper(grid, x, y+1, currentSum);
	//column
	else if(y == gridSize-1)
		return currentSum + find_min_path_helper(grid, x+1, y, currentSum);
	else
		return myMin(currentSum + find_min_path_helper(grid, x, y+1, currentSum),currentSum + find_min_path_helper(grid, x+1, y, currentSum));
}

unsigned int myMin(unsigned int num1, unsigned int num2)
{
	return num1>num2?num2:num1;
}

File: problem81.rb
def min(x,y)
	if x < y
		return x.to_i
	else
		return y.to_i
	end
end
 
matrix = Array.new(80)
i = 0

matrixFile = File.open("p081_matrix.csv", "r")
	matrixFile.each_line do |line|
		matrix[i] = Array.new(80)
		temp = line.split(',')
		j = 0
		while j < 80
			matrix[i][j] = temp[j].to_s.chomp.to_i
			j += 1
		end
		i += 1
	end
matrixFile.close()

i = 79 #row
j = 79 #column

while i > 0
	i -= 1
	matrix[i][j] += matrix[i+1][j]
end
i = 79
while j > 0
	j -= 1
	matrix[i][j] += matrix[i][j+1]
end

counter = 78

while counter > 0
	i = counter
	j = counter
	matrix[i][j] += min(matrix[i+1][j],matrix[i][j+1])

	while i > 0
		i -= 1
		matrix[i][j] += min(matrix[i+1][j],matrix[i][j+1])
	end
	i = counter
	while j > 0
		j -= 1
		matrix[i][j] += min(matrix[i+1][j],matrix[i][j+1])
	end
	
	counter -= 1
end

puts "Solution: #{matrix[0][0] + min(matrix[0][1],matrix[1][0])}"
